File: backend/src/main/java/com/smartrecruit/backend/service/CandidateService.java
package com.smartrecruit.backend.service;

import com.smartrecruit.backend.dto.candidate.CandidateRequest;
import com.smartrecruit.backend.dto.candidate.CandidateResponse;
import com.smartrecruit.backend.dto.candidate.CVSummaryResponse;
import com.smartrecruit.backend.entity.Candidate;
import com.smartrecruit.backend.entity.CV;
import com.smartrecruit.backend.entity.User;
import com.smartrecruit.backend.enums.RoleType;
import com.smartrecruit.backend.repository.CandidateRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CandidateService {

    private final CandidateRepository candidateRepository;

    public List<CandidateResponse> findAll(User currentUser) {
        ensureRecruiterOrAdmin(currentUser);
        return candidateRepository.findAll().stream()
                .map(this::toResponseWithoutCvs)
                .collect(Collectors.toList());
    }

    public CandidateResponse getById(UUID candidateId, User currentUser) {
        ensureRecruiterOrAdmin(currentUser);
        Candidate candidate = candidateRepository.findByIdWithCvs(candidateId)
                .orElseThrow(() -> new IllegalArgumentException("Candidate not found: " + candidateId));
        return toResponseWithCvs(candidate);
    }

    @Transactional
    public CandidateResponse create(CandidateRequest request, User currentUser) {
        ensureRecruiterOrAdmin(currentUser);
        Candidate candidate = Candidate.builder()
                .fullName(request.getFullName())
                .email(request.getEmail())
                .phone(request.getPhone())
                .build();
        candidate = candidateRepository.save(candidate);
        return toResponseWithoutCvs(candidate);
    }

    @Transactional
    public CandidateResponse update(UUID candidateId, CandidateRequest request, User currentUser) {
        ensureRecruiterOrAdmin(currentUser);
        Candidate candidate = candidateRepository.findById(candidateId)
                .orElseThrow(() -> new IllegalArgumentException("Candidate not found: " + candidateId));
        if (request.getFullName() != null) candidate.setFullName(request.getFullName());
        if (request.getEmail() != null) candidate.setEmail(request.getEmail());
        if (request.getPhone() != null) candidate.setPhone(request.getPhone());
        candidate = candidateRepository.save(candidate);
        return toResponseWithoutCvs(candidate);
    }

    @Transactional
    public void delete(UUID candidateId, User currentUser) {
        ensureRecruiterOrAdmin(currentUser);
        if (!candidateRepository.existsById(candidateId)) {
            throw new IllegalArgumentException("Candidate not found: " + candidateId);
        }
        candidateRepository.deleteById(candidateId);
    }

    private void ensureRecruiterOrAdmin(User user) {
        if (user.getRole() != RoleType.RECRUITER && user.getRole() != RoleType.ADMIN) {
            throw new org.springframework.security.access.AccessDeniedException("Only RECRUITER or ADMIN can manage candidates");
        }
    }

    private CandidateResponse toResponseWithoutCvs(Candidate c) {
        return CandidateResponse.builder()
                .id(c.getId())
                .fullName(c.getFullName())
                .email(c.getEmail())
                .phone(c.getPhone())
                .createdAt(c.getCreatedAt())
                .cvs(null)
                .build();
    }

    private CandidateResponse toResponseWithCvs(Candidate c) {
        List<CVSummaryResponse> cvs = c.getCvList().stream()
                .map(cv -> CVSummaryResponse.builder()
                        .id(cv.getId())
                        .filePath(cv.getFilePath())
                        .uploadedAt(cv.getUploadedAt())
                        .hasFeatures(cv.getFeatures() != null)
                        .build())
                .collect(Collectors.toList());
        return CandidateResponse.builder()
                .id(c.getId())
                .fullName(c.getFullName())
                .email(c.getEmail())
                .phone(c.getPhone())
                .createdAt(c.getCreatedAt())
                .cvs(cvs)
                .build();
    }
}

File: frontend/src/services/candidateService.ts
import apiClient from "./apiClient";

export interface Candidate {
  id: string;
  fullName: string;
  email: string;
  phone?: string;
  createdAt: string;
  updatedAt: string;
  recruiterId: string;
}

export interface CandidateRequest {
  fullName: string;
  email: string;
  phone?: string;
}

export interface CV {
  id: string;
  candidateId: string;
  filePath: string;
  fileName: string;
  uploadedAt: string;
  parsedText?: string;
}

export interface CVSummary {
  id: string;
  fileName: string;
  uploadedAt: string;
}

const candidateService = {
  getAll: async (): Promise<Candidate[]> => {
    const response = await apiClient.get("/candidates");
    return response.data;
  },

  getById: async (id: string): Promise<Candidate> => {
    const response = await apiClient.get(`/candidates/${id}`);
    return response.data;
  },

  create: async (data: CandidateRequest): Promise<Candidate> => {
    const response = await apiClient.post("/candidates", data);
    return response.data;
  },

  update: async (id: string, data: CandidateRequest): Promise<Candidate> => {
    const response = await apiClient.put(`/candidates/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/candidates/${id}`);
  },

  uploadCV: async (candidateId: string, file: File): Promise<CV> => {
    const formData = new FormData();
    formData.append("file", file);

    const response = await apiClient.post(`/candidates/${candidateId}/cvs`, formData);
    return response.data;
  },

  getCVs: async (candidateId: string): Promise<CVSummary[]> => {
    const response = await apiClient.get(`/candidates/${candidateId}/cvs`);
    return response.data;
  },

  getCV: async (candidateId: string, cvId: string): Promise<CV> => {
    const response = await apiClient.get(`/candidates/${candidateId}/cvs/${cvId}`);
    return response.data;
  },

  deleteCV: async (candidateId: string, cvId: string): Promise<void> => {
    await apiClient.delete(`/candidates/${candidateId}/cvs/${cvId}`);
  },
};

export default candidateService;
